src/coreclr/tools/Common/TypeSystem/IL/Stubs/AsyncThunks.cs

```diff
@@ -154,38 +159,22 @@ public static MethodIL EmitTaskReturningThunk(MethodDesc taskReturningMethod, Me
                     codestream.BeginHandler(tryCatchRegion);
 
                     MethodDesc fromExceptionMd;
-                    if (logicalReturnType != null)
+                    if (isValueTask)
                     {
-                        // Generate: returnType.FromException<T>(Exception)
-                        if (isValueTask)
-                        {
-                            fromExceptionMd = context.SystemModule
-                                .GetKnownType("System.Threading.Tasks"u8, "ValueTask"u8)
-                                .GetKnownMethod("FromException"u8, null);
-                        }
-                        else
-                        {
-                            fromExceptionMd = context.SystemModule
-                                .GetKnownType("System.Threading.Tasks"u8, "Task"u8)
-                                .GetKnownMethod("FromException"u8, null);
-                        }
-                        fromExceptionMd = fromExceptionMd.MakeInstantiatedMethod(new Instantiation(logicalReturnType));
+                        fromExceptionMd = context.SystemModule
+                            .GetKnownType("System.Threading.Tasks"u8, "ValueTask"u8)
+                            .GetKnownMethod("FromException"u8, null);
                     }
                     else
                     {
-                        // Generate: returnType.FromException(Exception)
-                        if (isValueTask)
-                        {
-                            fromExceptionMd = context.SystemModule
-                                .GetKnownType("System.Threading.Tasks"u8, "ValueTask"u8)
-                                .GetKnownMethod("FromException"u8, null);
-                        }
-                        else
-                        {
-                            fromExceptionMd = context.SystemModule
-                                .GetKnownType("System.Threading.Tasks"u8, "Task"u8)
-                                .GetKnownMethod("FromException"u8, null);
-                        }
```

Comment:
The refactoring here is not correct. The original code looked refactorable because it did the same thing, but that was a bug.

The `GetKnownMethod` needs to be called with a non-null signature here. These methods have two overloads and we need to call the correct one. This will need to be something along the lines `new MethodSignature(MethodSignatureFlags.Static, 1, ((MetadataType)returnType.GetTypeDefinition()).MakeInstantiatedType(context.GetSignatureVariable(0, true)), [])` for the generic one and similar for the non-generic one.